### PR TITLE
Flip sign convention for FFTs used by KSpace methods

### PR DESCRIPTION
**Pull request by sjplimp**
**Summary**

Flip the sign used in the LAMMPS FFTs to match the backward/forward convention use by other libs.
Decades ago this was more ambiguous, now standardized.  For the callers (PPPM methods) the only substantive change is that for intermediate results (convolution), the sign of the imaginary term is flipped.

**Related Issue(s)**

N/A

**Author(s)**

Steve and Stan

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Should be no changes if done correctly.

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by sjplimp** [comment]
Need to see if all regression tests that use PPPM methods past.  I don't think anything in USER-OMP, USER-INTEL, GPU is affected.  Possibly Kokkos is affected.  I've done several tests of PPPM and its options.  Less confident in pppm/dipole and pppm/disp at this point.

Also, this is being done in 2 stages.  First stage is changes to the caller classes.  Second stage will be to flip the convention in fft3d.cpp and one trivial change to the caller classes (flip an enum).

**Comment by akohlmey** [comment]
@sjplimp regression tests have completed (but they don't cover as much ground), but pppm/disp and pppm/disp/tip4p failed in the unittests: https://ci.lammps.org/job/dev/job/pull_requests/job/ubuntu/job/unit_tests/job/cmake_mpi_openmp_bigbig_shared/1432/
pppm/dipole is currently not being tested.

**Comment by sjplimp** [comment]
@akohlmey ok - see if this version of pppm/disp passes.  Do the tests also run examples/balance/in.balance.kspace, which uses pppm/disp ?

File: src/KSPACE/pppm.cpp [hunk 1]
```diff
@@ -51,6 +51,7 @@ using namespace MathSpecial;
 
 enum{REVERSE_RHO};
 enum{FORWARD_IK,FORWARD_AD,FORWARD_IK_PERATOM,FORWARD_AD_PERATOM};
+enum{FORWARD=-1,BACKWARD=1};
```
Comment (akohlmey) [inline src/KSPACE/pppm.cpp:54]: This enumerator should be placed inside the `FFT3d` class definition in `fft3d_wrap.h` and then referenced with `FFT3d::FORWARD` or `FFT3d::BACKWARD`.

**Comment by akohlmey** [comment]
> @akohlmey ok - see if this version of pppm/disp passes. Do the tests also run examples/balance/in.balance.kspace, which uses pppm/disp ?

nope. no change.

please note that also `src/USER-PHONON/fix_phonon. cpp` uses the `FFT3d` class and thus will need to be updated.

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 ok - I think pppm/disp and pppm/dipole should now be correct.  There were some bugs in what I had done in this PR, some bugs in the previous refactor for GridComm, and a couple long-standing bugs in pppm/disp.  Hopefully it will now pass regression tests.  @stanmoore1 is there anything to check/change in Kokkos?

**Comment by sjplimp** [comment]
@akohlmey did all the regression tests pass?  If so I will check in the final (mostly cosmetic) changes

**Comment by akohlmey** [comment]
> @akohlmey did all the regression tests pass? If so I will check in the final (mostly cosmetic) changes

looks like it.

**Comment by sjplimp** [comment]
@stanmoore1  I made changes to pppm/kokkos and KOKKOS fft3d.cpp/h that I thought were needed.  Please check if you agree and test them. @akohlmey this push does the internal flip of the FFT signs inside FFT3d.  I think I've addressed all your points above.  If this passes regression, I only have some documentation info to add.

**Comment by akohlmey** [comment]
> @stanmoore1 I made changes to pppm/kokkos and KOKKOS fft3d.cpp/h that I thought were needed. Please check if you agree and test them. @akohlmey this push does the internal flip of the FFT signs inside FFT3d. I think I've addressed all your points above. If this passes regression, I only have some documentation info to add.

compilation with KOKKOS fails. see: https://ci.lammps.org/job/dev/job/pull_requests/job/fedora/job/cmake_kokkos_mpi_openmp_clang_shared/1657/console
unittests pass since they do not test KOKKOS styles (yet).

**Comment by stanmoore1** [comment]
I'm working on the Kokkos porting now.

**Comment by stanmoore1** [comment]
I'm seeing a ~30% performance regression with this PR on V100 for rhodopsin 32K atoms. @weinbe2 any ideas? Would a forward vs backward FFT be more expensive in cuFFT? Seems like just flipping the sign should have no effect on speed.

**Comment by sjplimp** [comment]
@stanmoore1 Any perf change with non-Kokkos, non-GPU runs for rhodo?

**Comment by stanmoore1** [comment]
@sjplimp I'm not seeing a slowdown for non-Kokkos.

**Comment by stanmoore1** [comment]
The Kokkos GPU regression seems to be in `comm` time; the `nvprof` profiles are basically the same for new vs old.

**Comment by stanmoore1** [comment]
No slowdown for Kokkos OMP either, only GPU.

**Comment by stanmoore1** [comment]
Slowdown for Kokkos KISSFFT too, so not cuFFT specific

**Comment by stanmoore1** [comment]
I found the issue, I accidentally had a newer version of Kokkos installed, which had a performance regression (PEBCAK). I just need to check the per-atom for Kokkos, then this should be ready from my POV.

**Review by stanmoore1 — APPROVED** [review]

**Comment by stanmoore1** [comment]
Kokkos work is done.

**Pull request merged**